### PR TITLE
New driver for Brother wordprocessor/typewriter AX-145 [non-working]

### PR DESCRIPTION
**Pull request by BartmanAbyss**
split off [PR#8202](https://github.com/mamedev/mame/pull/8202)

### CODE REVIEW

File: src/mame/brother/ax145.cpp [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+
+	// bit 2: R/~W (Read / Not Write)
+	// bit 1: RS (Register Select)
+	// bit 0: E (Enable)
+	uint8_t lcd_signal{};
+
+	// driver_device overrides
+	void machine_start() override;
+	void machine_reset() override;
+	void video_start() override;
```
Comment (cuavas) [inline src/mame/brother/ax145.cpp:129]: Please make these `protected` like they are in the base class they’re inherited from.

File: src/mame/brother/ax145.cpp [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+private:
+	// devices
+	required_device<hd64180rp_device> maincpu;
+	required_device<hd44780_device> lcdc;
+	memory_share_creator<uint8_t> ram;
+	required_region_ptr<uint8_t> rom;
+
+	// valid values (bei IO3000=0x0b,0x07) (read_config @ 0x14c87): 0 = german => 0, 1 = german => 1, 2 = espanol => 2, 4 (gehäusedeckel offen) => 3, 8 = francais => 4, 16 = german => 5
+	static constexpr uint8_t id = 1;
```
Comment (cuavas) [inline src/mame/brother/ax145.cpp:116]: Please use `SCREAMING_SNAKE_CASE` for constants.

File: src/mame/brother/ax145.cpp [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+			if(BIT(lcd_signal, 1)) // RS
+				lcdc->data_w(data << 4);
+			else
+				lcdc->control_w(data << 4);
+		}
+	}
+
+	void dictionary_bank_w(uint8_t data) {
+		membank("dictionary")->set_entry(data & 0x03);
+	}
```
Comment (cuavas) [inline src/mame/brother/ax145.cpp:205]: This is a tag map lookup on the hot path.  Please use an object finder to find the memory bank.

File: src/mame/brother/ax145.cpp [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+		// config lower 4 bits
+		return (~id) & 0x0f;
+	}
+
+	uint8_t io_4000_r() {
+		//space.device().logerror("%s: IO 3000=%02x\n", pc(), io_3000);
+
+		// config upper 4 bits
+		return (~id) >> 4;
+	}
```
Comment (cuavas) [inline src/mame/brother/ax145.cpp:182]: Please use `logmacro.h` helpers for configurable logging.

File: src/mame/brother/ax145.cpp [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+	// int2
+	TIMER_DEVICE_CALLBACK_MEMBER(int2_timer_callback) {
+		maincpu->set_input_line(INPUT_LINE_IRQ2, ASSERT_LINE);
+	}
+
+	uint8_t irq_ack_r() {
+		maincpu->set_input_line(INPUT_LINE_IRQ2, CLEAR_LINE);
+		return 0;
+	}
```
Comment (cuavas) [inline src/mame/brother/ax145.cpp:215]: You need to check `machine().side_effects_disabled()` to suppress side effects for debugger reads.

**Review by cuavas — CHANGES_REQUESTED** [review]
This should be easy to get ready to merge.

**Comment by cuavas** [comment]
This is failing validation because you need to either flag the system `MACHINE_NO_SOUND` if it has sound hardware that isn’t emulated or `MACHINE_NO_SOUND_HW` if the machine itself has no sound hardware.

**Comment by BartmanAbyss** [comment]
> This is failing validation because you need to either flag the system `MACHINE_NO_SOUND` if it has sound hardware that isn’t emulated or `MACHINE_NO_SOUND_HW` if the machine itself has no sound hardware.

thanks. my bad. didn't know about `mame -validate`.

**Pull request merged**